packages/cosmic-swingset/src/computron-counter.js

```diff
@@ -0,0 +1,115 @@
+// @ts-check
+
+import { assert } from '@endo/errors';
+import {
+  BeansPerBlockComputeLimit,
+  BeansPerVatCreation,
+  BeansPerXsnapComputron,
+} from './sim-params.js';
+
+const { hasOwn } = Object;
+
+/**
+ * @typedef {object} BeansPerUnit
+ * @property {bigint} blockComputeLimit
+ * @property {bigint} vatCreation
+ * @property {bigint} xsnapComputron
+ */
+
+/**
+ * Return a stateful run policy that supports two phases: first allow
+ * non-cleanup work (presumably deliveries) until an overrideable computron
+ * budget is exhausted, then (iff no work was done and at least one vat cleanup
+ * budget field is positive) a cleanup phase that allows cleanup work (and
+ * presumably nothing else) until one of those fields is exhausted.
+ * https://github.com/Agoric/agoric-sdk/issues/8928#issuecomment-2053357870
+ *
+ * @param {object} params
+ * @param {BeansPerUnit} params.beansPerUnit
+ * @param {import('@agoric/swingset-vat').CleanupBudget} [params.vatCleanupBudget]
+ * @param {boolean} [ignoreBlockLimit]
+ * @returns {import('./launch-chain.js').ChainRunPolicy}
+ */
+export function computronCounter(
+  { beansPerUnit, vatCleanupBudget },
+  ignoreBlockLimit = false,
+) {
+  const {
+    [BeansPerBlockComputeLimit]: blockComputeLimit,
+    [BeansPerVatCreation]: vatCreation,
+    [BeansPerXsnapComputron]: xsnapComputron,
+  } = beansPerUnit;
+  assert.typeof(blockComputeLimit, 'bigint');
+  assert.typeof(vatCreation, 'bigint');
+  assert.typeof(xsnapComputron, 'bigint');
+
+  let totalBeans = 0n;
+  const shouldRun = () => ignoreBlockLimit || totalBeans < blockComputeLimit;
+
+  const remainingCleanups = { default: Infinity, ...vatCleanupBudget };
+  const defaultCleanupBudget = remainingCleanups.default;
+  let cleanupStarted = false;
+  let cleanupDone = false;
+  const cleanupPossible =
+    Object.values(remainingCleanups).length > 0
+      ? Object.values(remainingCleanups).some(n => n > 0)
+      : defaultCleanupBudget > 0;
```

Comment:
Please feel free to clean up this function a bit.
```suggestion
  const cleanupPossible = Object.values(remainingCleanups).some(n => n > 0);
```